File: daemon/src/features/ledger/errand.ts
import { stat } from 'node:fs/promises'
import path from 'node:path'
import { Git } from '@daemon/utils/git'

/**
 * What an errand did to a checkout, at the only granularity there is one: an agent hands a
 * piece of work to Claude Code or to a shell, and the disk either side of it says which files
 * came back different. Nothing here can say which line was whose, and the wording everywhere
 * downstream is "as part of" rather than "wrote" for that reason.
 *
 * A mark is what git says about a path and what the file is: the status alone would miss an
 * errand that edited a file somebody had already left dirty, since it reads as modified both
 * times. Only the paths git already calls dirty are marked, so this costs a `git status` and
 * a `stat` per changed file rather than a walk of the tree.
 */
export type Marks = Map<string, string>

export async function marksOf(checkout: string): Promise<Marks> {
  const changes = await new Git(checkout).changes().catch(() => ({}))
  const marks: Marks = new Map()
  await Promise.all(
    Object.entries(changes).map(async ([file, change]) => {
      const found = await stat(path.join(checkout, file)).catch(() => null)
      const shape = found ? `${String(found.mtimeMs)}:${String(found.size)}` : 'gone'
      marks.set(file, `${change}:${shape}`)
    }),
  )
  return marks
}

/**
 * The paths the errand left different, sorted. A path that stopped being dirty is not one of
 * them: an errand that committed what somebody else had already changed would otherwise be
 * filed as having written every one of those files, which is exactly the over-reading a
 * coarse record invites.
 */
export function changedBetween(before: Marks, after: Marks): string[] {
  const changed: string[] = []
  for (const [file, mark] of after) if (before.get(file) !== mark) changed.push(file)
  return changed.sort()
}

File: daemon-go/internal/errand/errand.go
// Package errand is what an errand left different, for the ledger.
//
// The hands work on the real disk rather than through the app's door, so watching the checkout
// either side of one is the only way the app can find out what they did. The boundary is all it
// knows: it says which errand a file was part of, never which line was whose.
package errand

import (
	"io/fs"
	"path/filepath"
	"slices"
	"strconv"
	"strings"

	"github.com/broodmotherai/broodmother/daemon-go/internal/constants"
	"github.com/broodmotherai/broodmother/daemon-go/internal/doc"
)

// Marks is every file in a checkout and what it looked like, keyed by the path the ledger files
// acts under.
type Marks map[doc.Path]string

// skipped are the folders no errand's work is in, and walking them is the difference between a
// look that costs nothing and one that reads a `node_modules`.
var skipped = map[string]bool{
	".git": true, "node_modules": true, ".next": true, "dist": true, "build": true,
	constants.ReposDir: true,
}

// MarksOf is one look at a checkout: each file's size and mtime, which is enough to tell a file
// that changed from one that did not without reading any of them.
func MarksOf(checkout string) Marks {
	held := Marks{}
	if checkout == "" {
		return held
	}
	filepath.WalkDir(checkout, func(path string, entry fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		name := entry.Name()
		if entry.IsDir() {
			if path != checkout && skipped[name] {
				return filepath.SkipDir
			}
			return nil
		}
		info, err := entry.Info()
		if err != nil {
			return nil
		}
		relative, err := filepath.Rel(checkout, path)
		if err != nil {
			return nil
		}
		held[doc.Path(filepath.ToSlash(relative))] =
			strconv.FormatInt(info.Size(), 10) + ":" + strconv.FormatInt(info.ModTime().UnixNano(), 10)
		return nil
	})
	return held
}

// ChangedBetween is what differs across two looks: written, edited or taken away. Sorted, so one
// errand's acts land in an order somebody can read rather than a map's.
func ChangedBetween(before, after Marks) []doc.Path {
	changed := []doc.Path{}
	for path, mark := range after {
		if before[path] != mark {
			changed = append(changed, path)
		}
	}
	for path := range before {
		if _, still := after[path]; !still {
			changed = append(changed, path)
		}
	}
	sortPaths(changed)
	return changed
}

func sortPaths(paths []doc.Path) {
	slices.SortFunc(paths, func(a, b doc.Path) int { return strings.Compare(string(a), string(b)) })
}
